### PR TITLE
Fix PolymorphismBoxing for more complex uses of interfaces and effect handlers

### LINKED ISSUE #574: Declarations of interfaces with block parameters are incorrectly translated to core
In the test `examples/pos/if_control_effect.effekt`, the interface declaration
```effekt
interface Control {
    def ifelse[A](cond: Bool) { tcase: () => A } { fcase: () => A }: A
}
```
gets translated to the following in core:
```effekt-core
interface Control2839 {
  ifelse2846: [A2842](Bool379) => A2842
}
```
That is, block parameters are dropped.

### Full output
`effekt --backend js --ir-show core examples/pos/object/if_control_effect.effekt` results in the following core code:
```effekt-core
module if_control_effect

import effekt
import option
import list
import result
import exception
import array
import string
import ref



interface Control2839 {
  ifelse2846: [A2842](Bool379) => A2842
}

def cImpl2840 = new Control2839 {
  def ifelse2846[A2847](cond2848: Bool379){t2849}{f2850} = 
    if (cond2848) {
      t2849()
    } else {
      f2850()
    }
};
def main2841() =
  cImpl2840.ifelse[Unit374](infixEq71(1, 2), { () => 
    let v_r_28834011 = println1("true")
    
    v_r_28834011
  }, { () => 
    let v_r_28844012 = println1("false")
    
    v_r_28844012
  });
  cImpl2840.ifelse[Unit374](infixEq71(1, 1), { () => 
    let v_r_28874013 = println1("true")
    
    v_r_28874013
  }, { () => 
    let v_r_28884014 = println1("false")
    
    v_r_28884014
  })

```

### PR DESCRIPTION
**Pull request by marzipankaiser**
This fixes the program mentioned by @marvinborner in #568.

In order to do this, this also fixes #574 (thats just 257c07a, if we'd want to cherry-pick)

cc @phischu

### CODE REVIEW

**Comment by marzipankaiser** [comment]
This now just fails one previous test case, which seems to be related to #574 (this now erases the non-declared arguments when coercing in this test case).

File: examples/llvm/polymorphic_codata_polymorphically.effekt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+interface Eff[A] {
```
Comment (marvinborner) [inline examples/llvm/polymorphic_codata_polymorphically.effekt:1]: Note that this is actually just a (minor) minification of [pos/higher_rank_polymorphism.effekt](https://github.com/effekt-lang/effekt/blob/master/examples/pos/higher_rank_polymorphism.effekt). I'm not sure whether we need this additional test sample.

Comment (marzipankaiser) [reply examples/llvm/polymorphic_codata_polymorphically.effekt:1]: Ah, good point. I'll try if that also detects this change.

Comment (marzipankaiser) [reply examples/llvm/polymorphic_codata_polymorphically.effekt]: It does, so I removed the new test case. Thanks.

**Comment by marvinborner** [comment]
`examples/pos/parametrized.effekt` now fails with "Wrong number of type arguments". Previously it just segfaulted.

Stacktrace:

```
[error] java.lang.AssertionError: assertion failed: Wrong number of type arguments
[error] 	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] 	at effekt.core.Type$.instantiate(Type.scala:100)
[error] 	at effekt.core.Type$.inferType(Type.scala:166)
[error] 	at effekt.core.Stmt.<init>(Tree.scala:314)
[error] 	at effekt.core.Stmt$App.<init>(Tree.scala:291)
[error] 	at effekt.core.Stmt$App$.apply(Tree.scala:291)
[error] 	at effekt.core.PolymorphismBoxing$$anon$5.apply(PolymorphismBoxing.scala:551)
[error] 	at effekt.core.PolymorphismBoxing$$anon$5.apply(PolymorphismBoxing.scala:542)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:227)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:289)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:281)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:281)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:212)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:217)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:195)
[error] 	at effekt.core.PolymorphismBoxing$.transform$$anonfun$3(PolymorphismBoxing.scala:155)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:293)
[error] 	at effekt.core.PolymorphismBoxing$.transform(PolymorphismBoxing.scala:155)
[error] 	at effekt.core.PolymorphismBoxing$.$anonfun$4(PolymorphismBoxing.scala:148)
[error] 	at effekt.util.Timers.timed(Timer.scala:39)
[error] 	at effekt.util.Timers.timed$(Timer.scala:12)
[error] 	at effekt.context.Context.timed(Context.scala:40)
[error] 	at effekt.core.PolymorphismBoxing$.run(PolymorphismBoxing.scala:148)
[error] 	at effekt.core.PolymorphismBoxing$.run(PolymorphismBoxing.scala:144)
[error] 	at effekt.Phase$$anon$1.run$$anonfun$1(Phase.scala:51)
[error] 	at scala.Option.flatMap(Option.scala:283)
...
```

**Comment by marzipankaiser** [comment]
@marvinborner Well, I guess this is better than a segfault ;)
I will look into it.

**Pull request merged**